### PR TITLE
Add Windows Error Reporting plugin

### PR DESCRIPTION
**Pull request by Poeloe**
(DIS-974)

### CODE REVIEW

File: dissect/target/plugins/os/windows/wer.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                key = None
+
+        return record_fields, record_values
+
+    @staticmethod
+    def _collect_wer_metadata(metadata_xml_file: TargetPath) -> tuple[list[tuple[str, str]], dict[str, str]]:
+        """Parse data from a metadata .xml file linked to a .wer file"""
+        record_fields = []
+        record_values = {}
+        file = metadata_xml_file.open().read().decode("utf-16")
```
[the remark this answers is not in the document]
Comment (Poeloe) [reply dissect/target/plugins/os/windows/wer.py]: Without "-le' it works like a charm ;)

File: dissect/target/plugins/os/windows/wer.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+camel_case_pattern_2 = re.compile(r"([a-z0-9])([A-Z])")
+camel_case_pattern_3 = re.compile(r"(\w)[.\s](\w)")
+
+
+def _collect_wer_data(wer_file: Path) -> tuple[list[tuple[str, str]], dict[str, str]]:
+    """Parse data from a .wer file."""
+    record_values = {}
+    record_fields = []
+    key = None
+    for line in wer_file.read_text("utf-16").splitlines():
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/wer.py:23]:
```suggestion
    for line in wer_file.read_text("utf-16-be").splitlines():
```
Maybe explicitly make it `-be` to make it clear.

File: dissect/target/plugins/os/windows/wer.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        if match := re.match(re.compile(r"(.+)=(.+)"), line.rstrip()):
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/wer.py:24]: Would a split on `=` also work?

File: dissect/target/plugins/os/windows/wer.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            key = None
+
+    return record_fields, record_values
+
+
+def _collect_wer_metadata(metadata_xml_file: Path) -> tuple[list[tuple[str, str]], dict[str, str]]:
+    """Parse data from a metadata .xml file linked to a .wer file."""
+    record_fields = []
+    record_values = {}
+    file = metadata_xml_file.read_text("utf-16")
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/wer.py:58]:
```suggestion
    file = metadata_xml_file.read_text("utf-16-be")
```
Same reason.

File: dissect/target/plugins/os/windows/wer.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            ("path", "metadata_file_path"),
+        ]
+    )
+    return TargetRecordDescriptor(record_name, record_fields)
+
+
+def _key_to_snake_case(key: str) -> str:
+    res = camel_case_pattern.sub(r"\1_\2", key)
+    res = camel_case_pattern_2.sub(r"\1_\2", res)
+    return camel_case_pattern_3.sub(r"\1_\2", res).lower()
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/wer.py:85]:
```suggestion
def _key_to_snake_case(key: str) -> str:
    for pattern in camel_case_patterns:
        key = pattern.sub(r"\1_\2", key)
    return key.lower()
```
Would this work? With this of course as well:

```python
camel_case_patterns = [
    re.compile(r"(\S)([A-Z][a-z]+)")
    re.compile(r"([a-z0-9])([A-Z])")
    re.compile(r"(\w)[.\s](\w)")
]
```

File: dissect/target/plugins/os/windows/wer.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            friendly_event_name: Human readable event name.
+            app_name: WER file application name.
+            app_path: Path to application that caused/reported the error.
+            report_description: WER file report description.
+            application_identity: WER file application identity.
+            metadata_hash: WER file metadata hash.
+        """
+        for files in self.wer_files:
+            wer_fields = []
+            wer_values = {"_target": self.target}
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/wer.py:148]:
```suggestion
            record_values = {"_target": self.target}
```
I think `record_values` is a bit more accurate.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/wer.py (hunk 2 above)
Comment (Poeloe) [reply dissect/target/plugins/os/windows/wer.py:21]: Talked about this one-on-one: files use a BOM, so we leave this as is.

File: dissect/target/plugins/os/windows/wer.py (hunk 3 above)
Comment (Poeloe) [reply dissect/target/plugins/os/windows/wer.py]: Yep, this unnecessarily complicated. Changed it!

File: dissect/target/plugins/os/windows/wer.py (hunk 5 above)
Comment (Poeloe) [reply dissect/target/plugins/os/windows/wer.py]: Yep, more elegant :)

**Review by Schamper — APPROVED** [review]

**Review by Schamper — APPROVED** [review]

**Pull request merged**